File: atak/ATAKMapEngine/lib/src/main/java/com/atakmap/map/layer/feature/geometry/opengl/GLGeometryCollection.java
package com.atakmap.map.layer.feature.geometry.opengl;

import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

import com.atakmap.annotations.DeprecatedApi;
import com.atakmap.map.layer.feature.geometry.Geometry;
import com.atakmap.map.layer.feature.geometry.GeometryCollection;

/** @deprecated use the batch feature renderering framework */
@Deprecated
@DeprecatedApi(since = "4.1", forRemoval = true, removeAt = "4.4")
public final class GLGeometryCollection extends GLGeometry {

    private List<GLGeometry> geometries;

    public GLGeometryCollection(GeometryCollection collection) {
        super(collection);
        
        this.geometries = new LinkedList<GLGeometry>();
        Collection<Geometry> c = collection.getGeometries();
        for(Geometry geom : c)
            this.geometries.add(GLGeometry.createRenderer(geom));
    }
    
    public Iterator<GLGeometry> iterator() {
        return this.geometries.iterator();
    }
}

File: takengine/mapengine/sdk/src/renderer/feature/GLGeometryCollection.cpp
#include "renderer/feature/GLGeometryCollection.h"

using namespace atakmap::feature;
using namespace atakmap::renderer::feature;

GLGeometryCollection::GLGeometryCollection(GeometryCollection *collection)
: GLGeometry(collection)
{
    std::pair<GeometryCollection::GeometryVector::const_iterator, GeometryCollection::GeometryVector::const_iterator> contents = collection->contents();
    size_t count = std::distance(contents.first, contents.second);
    this->geometries.reserve(count);
    while (contents.first != contents.second) {
        Geometry *geom = *contents.first;
        this->geometries.push_back(GLGeometry::createRenderer(geom));
        ++contents.first;
    }
}

GLGeometryCollection::~GLGeometryCollection() { }

GLGeometryCollection::GLGeometryIterator::GLGeometryIterator(GeometryList::iterator first,
                                                             GeometryList::iterator last)
: pos(first), end(last) { }

GLGeometryCollection::GLGeometryIterator::~GLGeometryIterator() { }

bool GLGeometryCollection::GLGeometryIterator::hasNext() {
    return this->pos != this->end;
}

GLGeometry *GLGeometryCollection::GLGeometryIterator::next() {
    ++pos;
    return *pos;
}

GLGeometry *GLGeometryCollection::GLGeometryIterator::get() {
    return *pos;
}

atakmap::port::Iterator<GLGeometry *> *GLGeometryCollection::getIterator() {
    return new GLGeometryIterator(this->geometries.begin(), this->geometries.end());
}
